packages/postgres-database/src/simcore_postgres_database/utils_folders.py

```diff
@@ -0,0 +1,597 @@
+                    ),
+                )
+            )
+        except ForeignKeyViolation as e:
+            raise GroupIdDoesNotExistError(gid=gid) from e
+
+        return _FolderID(folder_id)
+
+
+async def folder_share(
+    connection: SAConnection,
+    folder_id: _FolderID,
+    sharing_gids: set[_GroupID],
+    *,
+    recipient_gid: _GroupID,
+    parent_folder: _FolderID | None = None,
+    recipient_read: bool = False,
+    recipient_write: bool = False,
+    recipient_delete: bool = False,
+    recipient_admin: bool = False,
+) -> None:
+    # Permission change rules:
+    # - `owner`` can never loose any permission, always has all of them
+    # - `admin` can edit and remove permissions of every other user, including other `admins`
+    # - `admin` is granted only by the owner
+
+    async with connection.begin():
+        # check that owner permissions are never alterd not even by the owner itself
+        query_possible_owner_entry = await (
+            await connection.execute(
+                sa.select([folders, folders_access_rights])
+                .select_from(
+                    folders.join(
+                        folders_access_rights,
+                        folders.c.id == folders_access_rights.c.folder_id,
+                    )
+                )
+                .where(folders_access_rights.c.gid == recipient_gid)
+            )
+        ).fetchone()
+        if query_possible_owner_entry:
+            owner_entry = FolderEntry.from_orm(query_possible_owner_entry)
+            if owner_entry.owner == owner_entry.gid:
+                raise CannotAlterOwnerPermissionsError(
+                    gid=recipient_gid, folder_id=folder_id
+                )
+
+        if recipient_admin:
+            query_can_grant_admin = (
+                sa.select([folders, folders_access_rights])
+                .select_from(
+                    folders.join(
+                        folders_access_rights,
+                        folders.c.id == folders_access_rights.c.folder_id,
+                    )
+                )
+                .where(folders_access_rights.c.folder_id == folder_id)
+                .where(folders_access_rights.c.gid.in_(sharing_gids))
+                .where(folders_access_rights.c.gid == folders.c.owner)
+                .where(folders_access_rights.c.admin.is_(True))
+            )
+            can_grant_admin = await connection.scalar(query_can_grant_admin)
+            if not can_grant_admin:
+                raise RequiresOwnerToMakeAdminError(
+                    folder_id=folder_id, gids=sharing_gids
+                )
+
+        # check if any of the sharing groups have the required requested permissions
+        def _get_permissions_query(column: sa.Column) -> sa.sql.Select:
+            # NOTE: permissions can be granted if that group has the admin permission
+            # set and it also has access to the permission
+            return (
```

Comment:
I have a feeling that with this approach you can easily overload and kill DB, it is really unefficient to do so many calls to the DB (the same approach is in the Storage and now we have problems with it). I would do as less queries to DB as possible, I think you are overdoing it here a bit. Lets discuss in person please.